src/core/lib/gprpp/work_serializer.cc

```diff
@@ -253,14 +284,231 @@ void WorkSerializer::WorkSerializerImpl::DrainQueueOwned() {
+
+#ifndef NDEBUG
+  static thread_local DispatchingWorkSerializer* running_work_serializer_;
+#endif
+};
+
+#ifndef NDEBUG
+thread_local WorkSerializer::DispatchingWorkSerializer*
+    WorkSerializer::DispatchingWorkSerializer::running_work_serializer_ =
+        nullptr;
+#endif
+
+void WorkSerializer::DispatchingWorkSerializer::Orphan() {
+  ReleasableMutexLock lock(&mu_);
+  // If we're not running, then we can delete immediately.
+  if (!running_) {
+    lock.Release();
+    delete this;
+    return;
+  }
+  // Otherwise store a flag to delete when we're done.
+  orphaned_ = true;
+}
+
+// Implementation of WorkSerializerImpl::Run
+void WorkSerializer::DispatchingWorkSerializer::Run(
+    std::function<void()> callback, const DebugLocation& location) {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
+    gpr_log(GPR_INFO, "WorkSerializer[%p] Scheduling callback [%s:%d]", this,
+            location.file(), location.line());
+  }
+  MutexLock lock(&mu_);
+  if (!running_) {
+    // If we were previously idle, insert this callback directly into the empty
+    // processing_ list and start running.
+    running_ = true;
+    GPR_ASSERT(processing_.empty());
+    processing_.emplace_back(std::move(callback), location);
+    event_engine_->Run(this);
+  } else {
+    // We are already running, so add this callback to the incoming_ list.
+    // The work loop will eventually get to it.
+    incoming_.emplace_back(std::move(callback), location);
+  }
+}
+
+// Implementation of EventEngine::Closure::Run - our actual work loop
+void WorkSerializer::DispatchingWorkSerializer::Run() {
+  // TODO(ctiller): remove these when we can deprecate ExecCtx
+  ApplicationCallbackExecCtx app_exec_ctx;
+  ExecCtx exec_ctx;
+  // Grab the last element of processing_ - which is the next item in our queue
+  // since processing_ is stored in reverse order.
+  auto& cb = processing_.back();
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_work_serializer_trace)) {
+    gpr_log(GPR_INFO, "WorkSerializer[%p] Executing callback [%s:%d]", this,
+            cb.location.file(), cb.location.line());
+  }
+  // Run the work item.
+  SetCurrentThread();
+  cb.callback();
+  // pop_back here destroys the callback - freeing any resources it might hold.
+  // We do so before clearing the current thread in case the callback destructor
+  // wants to check that it's in the WorkSerializer too.
+  processing_.pop_back();
+  ClearCurrentThread();
+  // Check if we've drained the queue and if so refill it.
+  if (processing_.empty() && !Refill()) return;
+  // There's still work in processing_, so schedule ourselves again on
+  // EventEngine.
+  event_engine_->Run(this);
+}
+
+WorkSerializer::DispatchingWorkSerializer::RefillResult
+WorkSerializer::DispatchingWorkSerializer::RefillInner() {
+  // Recover any memory held by processing_, so that we don't grow forever.
+  // Do so before acquiring a lock so we don't cause inadvertent contention.
+  processing_.shrink_to_fit();
```

Comment:
Neither `resize(0)` nor `clear()` deallocate memory.

Reply:
Really?  From a glance at the implementation, it looks like [`clear()` does](https://github.com/abseil/abseil-cpp/blob/5655528c41830f733160de4fb0b99073841bae9e/absl/container/inlined_vector.h#L806).  It looks like [`resize()` might also](https://github.com/abseil/abseil-cpp/blob/5655528c41830f733160de4fb0b99073841bae9e/absl/container/inlined_vector.h#L602), although it's a bit hard to follow through all the templating.

Anyway, doesn't really matter, I was mainly curious.

Reply:
ah, forgot this was `absl::InlinedVector` - `std::vector` doesn't, and I prefer to keep things so we don't accidentally cause an outage should we change to `std::vector`. (It's an experiment I'd like to do in the future - it would significantly reduce idle memory usage)